sql: tests (and bugfix) for ? acceleration

jordanlewis opened the pull request:
This commit adds more unit testing for ? acceleration, and fixes a bug
that wasn't caught by existing tests: a ? search against an object that
contains a key that has the search term as a prefix was incorrectly
returning that object.

The bug is fixed by using a more precise span to search: instead of
searching for all keys with the search key as a prefix, search for the
span that starts with the search key and ends with PrefixEnd of the
search key plus the path separator. This includes only keys that are
precisely the search key (in the case of keys that point to empty
objects or arrays) or keys that don't end an object path (due to the
separator after them).

Release note: None (bug was not in released version)

jordanlewis committed:
sql: tests (and bugfix) for ? acceleration

This commit adds more unit testing for ? acceleration, and fixes a bug
that wasn't caught by existing tests: a ? search against an object that
contains a key that has the search term as a prefix was incorrectly
returning that object.

The bug is fixed by using a more precise span to search: instead of
searching for all keys with the search key as a prefix, search for the
span that starts with the search key and ends with PrefixEnd of the
search key plus the path separator. This includes only keys that are
precisely the search key (in the case of keys that point to empty
objects or arrays) or keys that don't end an object path (due to the
separator after them).

Release note: None (bug was not in released version)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81646)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Nice!

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-:-N2i2qR27Ydgj_HTfSuu:b-83sjoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[`pkg/util/json/json_test.go` line 1703 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i0M6F-UdG0ucsLyYB:-N2i0M6F-UdG0ucsLyYC:b-864w4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1703)):*
> ```Go
> 		indexedValue string
> 		value        string
> 		containsKeys bool
> ```

nit: I don't think you really need a separate value for `containsKeys` for this test, since it's always the same as `expected` (since `tight` is always true)
___
*[`pkg/util/json/json_test.go` line 1705 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i09kl1QF0pwSvUTOF:-N2i09kl1QF0pwSvUTOG:byjzuy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1705)):*
> ```Go
> 		containsKeys bool
> 		expected     bool
> 		unique       bool
> ```

I don't think you're using this `unique` value (you don't need it since it should always be false for exists)
___
*[`pkg/util/json/json_test.go` line 1713 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i-a6YBYODCl2PMYwu:-N2i-a6YBYODCl2PMYwv:b-qitilh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1713)):*
> not necessarily contained by the value.

copy pasta?
___
*[`pkg/util/json/json_test.go` line 1729 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i-jUB8F6kyHm4t0qJ:-N2i-jUCCJTDJ4YHI4ES:ba4j72q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1729)):*
> ```Go
> 		// Null doesn't have key string-null.
> 		{`null`, `null`, false, false, true},
> 		// Null doesn't have key string-null.
> ```

Null => []?
___
*[`pkg/util/json/json_test.go` line 1739 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i06MUDDT4uE_GYvSU:-N2i06MUDDT4uE_GYvSV:b-nhgycs) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1739)):*
> ```Go
> 	}
> 
> 	runTest := func(indexedValue JSON, value string, expectContainsKeys, expected, expectUnique bool) {
> ```

I don't think you're using the `expected` or `expectUnique` bools.
___
*[`pkg/util/json/json_test.go` line 1753 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i1G6A3pfVyKf7ofQz:-N2i1G6A3pfVyKf7ofR-:bsmpi1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1753)):*
> ```Go
> 		// Spans should always be tight for exists.
> 		if !spanExpr.Tight {
> 			t.Errorf("For %s, expected tight=false, but got true", value)
> ```

"tight=false, but got true" -> "tight=true, but got false"
___
*[`pkg/util/json/json_test.go` line 1799 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i29P2EzCh6po3zvXh:-N2i29P3AWS0ZefCdfrD:bd88mhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1799)):*
> ```Go
> 		require.NoError(t, err)
> 
> 		isContained, err := left.Exists(right)
> ```

nit: isContained -> exists


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81646)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/util/json/json_test.go` line 1705 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i09kl1QF0pwSvUTOF:-N2i9uvx6TjZIULvHlOc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1705)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I don't think you're using this `unique` value (you don't need it since it should always be false for exists)
</blockquote></details>

Done.
___
*[`pkg/util/json/json_test.go` line 1713 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i-a6YBYODCl2PMYwu:-N2i9uaxEcS_PRfguAWJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1713)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

> not necessarily contained by the value.

copy pasta?
</blockquote></details>

Done.
___
*[`pkg/util/json/json_test.go` line 1729 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i-jUB8F6kyHm4t0qJ:-N2i9uJvBquI3-TRUtOt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1729)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Null => []?
</blockquote></details>

Done.
___
*[`pkg/util/json/json_test.go` line 1739 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i06MUDDT4uE_GYvSU:-N2i9u0XCGhMNprNfCgC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1739)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I don't think you're using the `expected` or `expectUnique` bools.
</blockquote></details>

Done.
___
*[`pkg/util/json/json_test.go` line 1753 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-N2i1G6A3pfVyKf7ofQz:-N2i9tgHEzDGNZ4X1RBx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f6ffb528c4c76ffdae8edffb027240f710dc6c/pkg/util/json/json_test.go#L1753)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

"tight=false, but got true" -> "tight=true, but got false"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
This testing has found another bug in code that's existed since the initial 2017 implementation of the `pkg/util/json` package. `jsonEncoded.Exists` doesn't always return the same result as `jsonArray.Exists` does, for the same inputs. I'll fix that separately, but this won't be mergeable until I fix it because of the randomized tests in this PR.

jordanlewis committed:
json: bugfix to jsonArray.Exists

Previously, the JSON Exists method (the `?` operator in SQL) could
return false negatives if invoked on a JSON array that had been
shallowDecoded to a single level.

Release note (bug fix): fix false negatives produced by the JSON `?`
operator when invoked on a JSON array with the vectorized engine set
explicitly to off.

rytaft reviewed (approved):
:thumbsup: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81646#-:-N2iaqZvBMsSHWktXubM:b-3weqpr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
First commit is now #81648

jordanlewis commented:
bors r+
